### PR TITLE
API Rest+GraphQL 

### PR DESCRIPTION
**Pull request by fjnoyp**
This PR combines the Rest and GraphQL API components as one API module in Dart/native. 

**The REST API component of this PR has already been reviewed in this PR: https://github.com/aws-amplify/amplify-flutter/pull/261**
Please note that the Rest API PR comments have been addressed in this branch and not in the original Rest API branch from before. 

**The GraphQL API component of this PR has already been reviewed in this PR: https://github.com/aws-amplify/amplify-flutter/pull/259**

Code changes were made to the Rest and GraphQL components to harmonize the code implementations and ensure consistent error handling approaches. 

Cancellation support via storage of the Operation objects in native was introduced for GraphQL.  The unit tests for that functionality still needs to be added in the future. 

The dart example app has been merged to allow switching between a Rest API tester and a GraphQL tester. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**You might want to wait for Ashish to merge in his GraphQL PR before reviewing this PR to reduce the number of "files changed" to review**

### CODE REVIEW

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        LOG.info("Initiated API plugin")
+    }
+
+    override fun onMethodCall( call: MethodCall, result: Result) {
+        var methodName = call.method
+
+        if(methodName == "cancel"){
+          onCancel(result, (call.arguments as String))
+          return
+        }
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt:51]:
```suggestion
        if(methodName == "cancel") {
            onCancel(result, (call.arguments as String))
            return
        }
```

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                    FlutterApiErrorMessage.ERROR_CASTING_INPUT_IN_PLATFORM_CODE.toString(),
+                    e
+            )
+        }
+    }
+
+    fun onCancel(
+            flutterResult: Result,
+            cancelToken: String) {
+        if(OperationsManager.containsOperation(cancelToken)){
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt:79]:
```suggestion
        if(OperationsManager.containsOperation(cancelToken)) {
```

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    
+    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.amazonaws.amplify/api")
+        channel.setMethodCallHandler(this)
+        context = flutterPluginBinding.applicationContext
+        Amplify.addPlugin(AWSApiPlugin())
+        LOG.info("Initiated API plugin")
+    }
+
+    override fun onMethodCall( call: MethodCall, result: Result) {
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt:45]:
```suggestion
    override fun onMethodCall(call: MethodCall, result: Result) {
```

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiErrorUtils.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            ))
+            return errorMap
+        }
+
+        fun handleAPIError(
+                flutterResult: Result,
+                msg: String,
+                localizedMessage: String,
+                recoverySuggestion: String
+        ){
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiErrorUtils.kt:68]:
```suggestion
        fun handleAPIError(
                flutterResult: Result,
                msg: String,
                localizedMessage: String,
                recoverySuggestion: String
        ) {
```

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiRequestUtils.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                var arguments = args as Map<String, Any>
+                var restOptions = arguments["restOptions"] as Map<String, Any>
+                var cancelToken = arguments["cancelToken"] as String
+            } catch (e: ClassCastException) {
+                FlutterApiErrorUtils.postFlutterError(
+                        flutterResult,
+                        FlutterApiErrorMessage.ERROR_CASTING_INPUT_IN_PLATFORM_CODE.toString(),
+                        e
+                )
+                return false;
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiRequestUtils.kt:37]: we shouldn't be posting errors from these helper/utils etc. You can throw exceptions here and catch in your caller to post flutter errors.

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                var cancelToken = arguments["cancelToken"] as String
+                var path = restOptions[PATH_KEY] as String
+
+            } catch (e: ClassCastException) {
+                FlutterApiErrorUtils.postFlutterError(
+                        flutterResult,
+                        FlutterApiErrorMessage.ERROR_CASTING_INPUT_IN_PLATFORM_CODE.toString(),
+                        e
+                )
+                return false;
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt:52]: Same, we should avoid posting errors from the validators

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+package com.amazonaws.amplify.amplify_api.rest_api
+
+import com.amazonaws.amplify.amplify_api.FlutterApiErrorMessage
+import com.amazonaws.amplify.amplify_api.FlutterApiErrorUtils
+import com.amplifyframework.api.rest.RestOptions
+import java.lang.ClassCastException
+import io.flutter.plugin.common.MethodChannel.Result
+
+
+data class FlutterRestInputs(val flutterInputsMap: Map<String, *>) {
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt:25]: suggestion: s/FlutterRestInputs/AmplifyAPIRestInputs/

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        private val BODY_KEY = "body"
+        private val QUERY_PARAM_KEY = "queryParameters"
+        private val HEADERS_KEY = "headers"
+
+        fun isValid(flutterResult: Result, args: Map<String, *>?): Boolean {
+            try {
+                var arguments = args as Map<String, *>
+
+                var restOptions = args[REST_OPTIONS_KEY] as Map<String, *>
+                var cancelToken = arguments["cancelToken"] as String
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt:43]:
```suggestion
                var cancelToken = arguments[CANCEL_TOKEN_KEY] as String
```

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestResponse.kt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ * permissions and limitations under the License.
+ */
+
+package com.amazonaws.amplify.amplify_api.rest_api
+
+import com.amplifyframework.api.rest.RestResponse
+
+import com.google.gson.Gson
+
+data class FlutterRestResponse(private var raw: RestResponse) {
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestResponse.kt:22]: suggestion: s/FlutterRestResponse/AmplifyAPIRestResponse/ or s/FlutterRestResponse/RestAPIResponse/

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/RestApiModule.kt [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+import com.amplifyframework.api.rest.RestOperation
+import com.amplifyframework.api.rest.RestOptions
+import com.amplifyframework.api.rest.RestResponse
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.Consumer
+import io.flutter.plugin.common.MethodChannel.Result
+import kotlin.reflect.KFunction3
+import kotlin.reflect.KFunction4
+
+typealias RestAPIFunction3 = KFunction3<
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/RestApiModule.kt:18]: suggestion: s/RestAPIFunction3/FunctionWithoutApiName/

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/RestApiModule.kt [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        @ParameterName(name = "restOptions") RestOptions,
+        @ParameterName(name = "restConsumer") Consumer<RestResponse>,
+        @ParameterName(name = "exceptionConsumer") Consumer<ApiException>,
+        RestOperation?>
+
+typealias RestAPIFunction4 = KFunction4<
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/RestApiModule.kt:24]: suggestion: s/RestAPIFunction3/FunctionWithApiName/

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/RestApiModule.kt [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+class RestApiModule {
+    companion object {
+        private val LOG = Amplify.Logging.forNamespace("amplify:flutter:api")
+
+        private fun restFunctionHelper(
+                methodName: String,
+                flutterResult: Result,
+                request: Map<String, *>,
+                function3: RestAPIFunction3,
+                function4: RestAPIFunction4
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/RestApiModule.kt:40]: suggestion: s/function3/functionWithoutApiName/

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt (hunk 7 above)
Comment (fjnoyp) [reply packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/rest_api/FlutterRestInputs.kt:25]: I've removed this class and merged it with FlutterApiRequest

**Pull request closed without merging**